lightning/src/ln/channelmanager.rs

```diff
@@ -7426,6 +7639,214 @@ where
 		}
 	}
 
+	#[cfg(dual_funding)]
+	fn internal_tx_add_input(&self, counterparty_node_id: &PublicKey, msg: &msgs::TxAddInput) -> Result<(), MsgHandleErrInternal> {
+		let per_peer_state = self.per_peer_state.read().unwrap();
+		let peer_state_mutex = per_peer_state.get(counterparty_node_id)
+			.ok_or_else(|| {
+				debug_assert!(false);
+				MsgHandleErrInternal::send_err_msg_no_close(
+					format!("Can't find a peer matching the passed counterparty node_id {}", counterparty_node_id),
+					msg.channel_id)
+			})?;
+		let mut peer_state_lock = peer_state_mutex.lock().unwrap();
+		let peer_state = &mut *peer_state_lock;
+		match peer_state.channel_by_id.entry(msg.channel_id) {
+			hash_map::Entry::Occupied(mut chan_phase_entry) => {
+				let channel_phase = chan_phase_entry.get_mut();
+				let msg_send_event = match channel_phase {
+					ChannelPhase::UnfundedInboundV2(ref mut channel) => {
+						channel.tx_add_input(msg).into_msg_send_event(counterparty_node_id)
+					},
+					ChannelPhase::UnfundedOutboundV2(ref mut channel) => {
+						channel.tx_add_input(msg).into_msg_send_event(counterparty_node_id)
+					},
+					_ => try_chan_phase_entry!(self, Err(ChannelError::Warn(
+						"Got a tx_add_input message with no interactive transaction construction expected or in-progress"
+						.into())), chan_phase_entry)
```

Comment:
Related to my other comment about putting the `InteractiveTxConstructor` in the relevant phase variants, would that remove the need for the `InteractivelyFunded` trait? Seems the check here is essentially being repeated there. And here the context is available so you don't need `HasChannelContext` if all the code in `InteractivelyFunded` were inlined in these methods. Just feels like this could be simplified.

Reply:
Yeah, those methods were inlined in the respective variants in the previous iteration. We had quite a bit of duplication before. I'll look into simplifying without re-introducing the duplication.

Reply:
I was thinking something like this:

```diff
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 1f65d064d..185daec8d 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -1243,6 +1243,19 @@ impl<'a, SP: Deref> ChannelPhase<SP> where
                        ChannelPhase::UnfundedInboundV2(ref mut chan) => &mut chan.context,
                }
        }
+
+       #[cfg(any(dual_funding, splicing))]
+       pub fn interactive_tx_constructor_mut(&'a mut self) -> Result<Option<&'a mut InteractiveTxConstructor>, ()> {
+               match self {
+                       ChannelPhase::Funded(ref mut chan) => Ok(chan.interactive_tx_constructor.as_mut()),
+                       ChannelPhase::UnfundedOutboundV1(ref mut chan) => Err(()),
+                       ChannelPhase::UnfundedInboundV1(ref mut chan) => Err(()),
+                       #[cfg(any(dual_funding, splicing))]
+                       ChannelPhase::UnfundedOutboundV2(ref mut chan) => Ok(chan.interactive_tx_constructor.as_mut()),
+                       #[cfg(any(dual_funding, splicing))]
+                       ChannelPhase::UnfundedInboundV2(ref mut chan) => Ok(chan.interactive_tx_constructor.as_mut()),
+               }
+       }
 }
 
 /// Contains all state common to unfunded inbound/outbound channels.
@@ -1652,17 +1665,6 @@ pub(super) trait InteractivelyFunded<SP: Deref> where SP::Target: SignerProvider
                Ok(msg)
        }
 
-       fn tx_add_input(&mut self, msg: &msgs::TxAddInput) -> InteractiveTxMessageSendResult {
-               InteractiveTxMessageSendResult(match self.interactive_tx_constructor_mut() {
-                       Some(ref mut tx_constructor) => tx_constructor.handle_tx_add_input(msg).map_err(
-                               |reason| reason.into_tx_abort_msg(self.context_mut().channel_id())),
-                       None => Err(msgs::TxAbort {
-                               channel_id: self.context_mut().channel_id(),
-                               data: "We do not have an interactive transaction negotiation in progress".to_string().into_bytes()
-                       }),
-               })
-       }
-
        fn tx_add_output(&mut self, msg: &msgs::TxAddOutput)-> InteractiveTxMessageSendResult {
                InteractiveTxMessageSendResult(match self.interactive_tx_constructor_mut() {
                        Some(ref mut tx_constructor) => tx_constructor.handle_tx_add_output(msg).map_err(
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 7f1000ab9..077e022cb 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -53,7 +53,7 @@ use crate::ln::features::{Bolt12InvoiceFeatures, ChannelFeatures, ChannelTypeFea
 #[cfg(any(feature = "_test_utils", test))]
 use crate::ln::features::Bolt11InvoiceFeatures;
 #[cfg(any(dual_funding, splicing))]
-use crate::ln::interactivetxs::InteractiveTxMessageSend;
+use crate::ln::interactivetxs::{InteractiveTxMessageSend, InteractiveTxMessageSendResult};
 use crate::routing::router::{BlindedTail, InFlightHtlcs, Path, Payee, PaymentParameters, Route, RouteParameters, Router};
 use crate::ln::onion_payment::{check_incoming_htlc_cltv, create_recv_pending_htlc_info, create_fwd_pending_htlc_info, decode_incoming_update_add_htlc_onion, InboundHTLCErr, NextPacketDetails};
 use crate::ln::msgs;
@@ -7864,18 +7864,23 @@ where
                match peer_state.channel_by_id.entry(msg.channel_id) {
                        hash_map::Entry::Occupied(mut chan_phase_entry) => {
                                let channel_phase = chan_phase_entry.get_mut();
-                               let msg_send_event = match channel_phase {
-                                       ChannelPhase::UnfundedInboundV2(ref mut channel) => {
-                                               channel.tx_add_input(msg).into_msg_send_event(counterparty_node_id)
-                                       },
-                                       ChannelPhase::UnfundedOutboundV2(ref mut channel) => {
-                                               channel.tx_add_input(msg).into_msg_send_event(counterparty_node_id)
+                               let msg_send_event = match channel_phase.interactive_tx_constructor_mut() {
+                                       Ok(Some(tx_constructor)) => {
+                                               tx_constructor
+                                                       .handle_tx_add_input(msg)
+                                                       .map_err(|reason| reason.into_tx_abort_msg(channel_phase.context_mut().channel_id()))
                                        },
-                                       _ => try_chan_phase_entry!(self, Err(ChannelError::Warn(
+                                       Ok(None) => Err(msgs::TxAbort {
+                                               channel_id: channel_phase.context_mut().channel_id(),
+                                               data: "We do not have an interactive transaction negotiation in progress".to_string().into_bytes(),
+                                       }),
+                                       Err(()) => try_chan_phase_entry!(self, Err(ChannelError::Warn(
                                                "Got a tx_add_input message with no interactive transaction construction expected or in-progress"
                                                .into())), chan_phase_entry)
                                };
-                               peer_state.pending_msg_events.push(msg_send_event);
+                               peer_state.pending_msg_events.push(
+                                       InteractiveTxMessageSendResult(msg_send_event).into_msg_send_event(counterparty_node_id)
+                               );
                                Ok(())
                        },
                        hash_map::Entry::Vacant(_) => {
```

Though I admit returning a `Result` in `ChannelPhase::interactive_tx_constructor_mut` might not be ideal. What you have is probably a bit safer, now that I think about it more. And it may be needed for `begin_interactive_funding_tx_construction`, anyway.